mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,277 @@
+use async_std::prelude::FutureExt;
+use chrono::Utc;
+use common::{log, HttpStatusCode, StatusCode};
+use derive_more::Display;
+use futures::channel::oneshot::{self, Receiver, Sender};
+use instant::Duration;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::MmError;
+use mm2_libp2p::{decode_message, encode_message, pub_sub_topic, Libp2pPublic, PeerId, TopicPrefix};
+use mm2_net::p2p::P2PContext;
+use ser_error_derive::SerializeErrorType;
+use serde::{Deserialize, Serialize};
+use std::str::FromStr;
+
+use crate::lp_network::broadcast_p2p_msg;
+
+pub(crate) const PEER_HEALTHCHECK_PREFIX: TopicPrefix = "hcheck";
+
+#[derive(Debug, Deserialize, Serialize)]
+#[cfg_attr(any(test, target_arch = "wasm32"), derive(PartialEq))]
+pub(crate) struct HealthcheckMessage {
+    signature: Vec<u8>,
+    data: HealthcheckData,
+}
+
+impl HealthcheckMessage {
+    pub(crate) fn generate_message(
+        ctx: &MmArc,
+        target_peer: PeerId,
+        is_a_reply: bool,
+        expires_in_seconds: i64,
+    ) -> Result<Self, String> {
+        let p2p_ctx = P2PContext::fetch_from_mm_arc(ctx);
+        let sender_peer = p2p_ctx.peer_id().to_string();
+        let keypair = p2p_ctx.keypair();
+        let sender_public_key = keypair.public().encode_protobuf();
+        let target_peer = target_peer.to_string();
+
+        let data = HealthcheckData {
+            sender_peer,
+            sender_public_key,
+            target_peer,
+            expires_at: Utc::now().timestamp() + expires_in_seconds,
+            is_a_reply,
+        };
+
+        let signature = try_s!(keypair.sign(&try_s!(data.encode())));
+
+        Ok(Self { signature, data })
+    }
+
+    pub(crate) fn is_received_message_valid(&self, my_peer_id: PeerId) -> bool {
+        let now = Utc::now().timestamp();
+        if now > self.data.expires_at {
+            log::debug!(
+                "Healthcheck message is expired. Current time in UTC: {now}, healthcheck `expires_at` in UTC: {}",
+                self.data.expires_at
+            );
+            return false;
+        }
+
+        if self.data.target_peer != my_peer_id.to_string() {
+            log::debug!(
+                "`target_peer` doesn't match with our peer address. Our address: '{}', healthcheck `target_peer`: '{}'.",
+                my_peer_id,
+                self.data.target_peer
+            );
+            return false;
+        }
+
+        let Ok(public_key) = Libp2pPublic::try_decode_protobuf(&self.data.sender_public_key) else {
+            log::debug!("Couldn't decode public key from the healthcheck message.");
+
+            return false
+        };
+
+        if self.data.sender_peer != public_key.to_peer_id().to_string() {
```

Comment:
isn't `sender_peer` redundant then? if it could be derived from the public key that's already in the `data` sent.

Reply:
IMO it's not redundant because we read it here:

https://github.com/KomodoPlatform/komodo-defi-framework/blob/14eeaeeb3da0745f7f4af1bf241e9c2d8755bc50/mm2src/mm2_main/src/lp_network.rs#L239-L256

right after receiving the message, which is a lot cheaper than deriving it from pubkey.